File: engine-dart/lib/src/core/middleware.dart
import 'package:shelf/shelf.dart' as shelf;

import '../types/context.dart';

/// Context available to middleware.
class MiddlewareContext {
  final shelf.Request request;
  final RequestInfo requestInfo;
  final String appId;
  final String path;
  final Map<String, dynamic> configuration;

  const MiddlewareContext({
    required this.request,
    required this.requestInfo,
    required this.appId,
    required this.path,
    this.configuration = const {},
  });
}

/// Response from onRequest middleware to short-circuit.
class MiddlewareResponse {
  final int status;
  final Map<String, String>? headers;
  final dynamic body;

  const MiddlewareResponse({required this.status, this.headers, this.body});
}

/// Middleware contract.
abstract class Middleware {
  String get name;
  Future<MiddlewareResponse?> onRequest(MiddlewareContext ctx) async => null;
  Future<shelf.Response> onResponse(
          MiddlewareContext ctx, shelf.Response response) async =>
      response;
}

File: engine/src/core/middleware.ts
import type { RequestInfo } from "../types/context";

// ── Middleware Types ──────────────────────────────────────────

export interface MiddlewareContext {
  request: Request;
  requestInfo: RequestInfo;
  appId: string;
  path: string;
  configuration: Record<string, unknown>;
}

export interface MiddlewareResponse {
  status: number;
  headers?: Record<string, string>;
  body?: unknown;
}

export interface Middleware {
  name: string;
  onRequest?: (context: MiddlewareContext) => Promise<MiddlewareResponse | void> | MiddlewareResponse | void;
  onResponse?: (context: MiddlewareContext, response: Response) => Promise<Response> | Response;
}
